hypothesis-python/src/hypothesis/internal/conjecture/engine.py

```diff
@@ -453,6 +453,14 @@ def test_function(self, data: ConjectureData) -> None:
                     ),
                 }
                 self.stats_per_test_case.append(call_stats)
+                if self.settings.backend != "hypothesis":
+                    for node in data.examples.ir_tree_nodes:
+                        value = data.provider.post_test_case_hook(node.value)
+                        assert (
+                            value is not None
+                        ), "providers must return a non-null value from post_test_case_hook"
```

Comment:
We could assert that `type(value) is expected_type` here, for a more rigorous check?

(note: not issubclass or our usual helper, because bool for an int node should be rejected)